Task Complete

Glasis9 opened the pull request:
![Снимок экрана 2023-02-14 170327](https://user-images.githubusercontent.com/108976083/218777272-3fb6c2c9-a477-457d-8ee8-dc9e0ba44d1d.jpg)

Glasis9 committed:
Task Complete, but I can’t make an async version, maybe the logic needs to be changed, what do you say?

Glasis9 committed:
fix

Glasis9 commented:
Task Complete, but I can’t make an async version, maybe the logic needs to be changed, what do you say?

ArsenPidhoretskyi commented on app/parse.py:
```diff
@@ -8,9 +16,108 @@ class Quote:
+
+
+def get_quotes() -> ([Quote], [Author]):
+    list_quote = []
```
rename to quotes

ArsenPidhoretskyi commented on app/parse.py:
```diff
@@ -8,9 +16,108 @@ class Quote:
+
+def get_quotes() -> ([Quote], [Author]):
+    list_quote = []
+    list_authors_url = []
```
authors_url

ArsenPidhoretskyi commented on app/parse.py:
```diff
@@ -8,9 +16,108 @@ class Quote:
+    return list_quote, list_author_bio
+
+
+def write_quotes_in_file(quotes: [Quote], output_csv_path: str) -> None:
```
[Quote]?
It maybe list[Quote]?
The same with Author

Glasis9 commented:
I'll fix all these little things, now I'm working on the processes

Glasis9 committed:
Rewrote the logic in parse.py (sync with author), in parser_class_processing unable to run code in parallel, multiple recursion error.
Tell me how to fix, please

olena-hul commented on app/parse_class_processing.py:
```diff
@@ -0,0 +1,155 @@
+
+
+@dataclass
+class Quote:
```
Lots of code duplicates in 3 files. Fix them

Glasis9 replied:
I know, fix this after fix parallel run code in parse_class_processin, ok?))

Glasis9 replied:
Can you see the code and help fix the errors?

ArsenPidhoretskyi commented on app/parse.py:
```diff
@@ -8,9 +18,102 @@ class Quote:
+def get_authors_page_soup(
+        page_soup: list[BeautifulSoup]
+) -> list[BeautifulSoup]:
+    list_authors = []
```
list_authors -> authors
Why you add prefix list. We know it already)

Glasis9 replied:
fix

ArsenPidhoretskyi commented on app/parse.py:
```diff
@@ -8,9 +18,102 @@ class Quote:
+        if url not in AUTHORS_URL_SET:
+            page = requests.get(url).content
+            list_authors.append(BeautifulSoup(page, "html.parser"))
+        AUTHORS_URL_SET.add(url)
```
Change constants???????
https://en.wikipedia.org/wiki/Constant_(computer_programming)

Glasis9 replied:
This is a global variable that stores links to authors in set to remove duplication.
I did not find another way to do this without classes and with the implementation of parallelism (so as not to pull data from processes)

ArsenPidhoretskyi replied:
It is constant. Then rename it to lowercase
https://intellipaat.com/blog/tutorial/python-tutorial/python-variables/#no3

ArsenPidhoretskyi commented on app/parse.py:
```diff
@@ -8,9 +18,102 @@ class Quote:
+
+
+def get_all_page_soup() -> list[BeautifulSoup]:
+    list_page_soup = []
```
remove prefix list_

ArsenPidhoretskyi replied:
The same in all places

Glasis9 replied:
fix

ArsenPidhoretskyi commented on app/parse_class.py:
```diff
@@ -0,0 +1,116 @@
+            ).text.replace("\n", " "),
+        )
+
+    @staticmethod
+    def _write_quotes_in_file(quotes: [Quote]) -> None:
+        quote_fields = [field.name for field in fields(Quote)]
+        with open("quotes.csv", "w", newline="", encoding="utf-8") as file:
+            writer = csv.writer(file)
+            writer.writerow(quote_fields)
+            writer.writerows([astuple(quote) for quote in quotes])
+
+    @staticmethod
+    def _write_authors_in_file(authors: [Author]) -> None:
+        author_fields = [field.name for field in fields(Author)]
+        with open("authors.csv", "w", newline="", encoding="utf-8") as file:
+            writer = csv.writer(file)
+            writer.writerow(author_fields)
+            writer.writerows([astuple(author) for author in authors])
```
As I can see only one change here
Create additional method and reuse it

ArsenPidhoretskyi commented on app/parse_class.py:
```diff
@@ -0,0 +1,116 @@
+        )
+
+    @staticmethod
+    def _write_quotes_in_file(quotes: [Quote]) -> None:
```
list[Quote]

ArsenPidhoretskyi commented on app/parse_class.py:
```diff
@@ -0,0 +1,116 @@
+            writer.writerows([astuple(quote) for quote in quotes])
+
+    @staticmethod
+    def _write_authors_in_file(authors: [Author]) -> None:
```
list[Author]

ArsenPidhoretskyi commented on app/parse_class.py:
```diff
@@ -0,0 +1,116 @@
+            self,
+            page_soup: list[BeautifulSoup]
+    ) -> list[BeautifulSoup]:
+        list_authors = []
```
remove prefix list_

ArsenPidhoretskyi replied:
The same in other places

ArsenPidhoretskyi commented on app/parse_class_processing.py:
```diff
@@ -0,0 +1,155 @@
+@dataclass
+class Author:
+    biography: str
+
```
Move classes to another file and reuse them
Do not repeat yourself!

ArsenPidhoretskyi reviewed (changes requested):
1. Why you always add prefixes or suffix with type list, set? It is too bad

Glasis9 committed:
Ok, I fix all your comments.
What should I do with parallel launch in parse_class_processing.py, your help me?

Glasis9 committed:
fix flake8

ArsenPidhoretskyi commented on app/parse.py:
```diff
@@ -1,4 +1,13 @@
+from urllib.parse import urljoin
+
+BASE_URL = "https://quotes.toscrape.com"
+AUTHORS_URL_SET = set()
```
Redundant suffix _SET

Glasis9 replied:
> Why you always add result_ to all names? Please explain this) I do not see why we need this

because it is a constant and even in the video Danil used upper case, if it's a bug, I'll fix it
I don't understand what this has to do with helping to implement a parallel launch.

Glasis9 replied:
so it is easier to separate the global variable from the non-global variable, in the attributes of the class I use lowercase letters

ArsenPidhoretskyi replied:
> because it is a constant and even in the video Danil used upper case,
As you say. Please read a little info documentation and check your code. 
https://intellipaat.com/blog/tutorial/python-tutorial/python-variables/#no3
https://en.wikipedia.org/wiki/Constant_(computer_programming)

ArsenPidhoretskyi reviewed (changes requested):
Why you always add result_ to all names?
Please explain this) I do not see why we need this

ArsenPidhoretskyi commented:
@Glasis9 If you have some questions about task better to ask them in chat no in the MR because next mentor will not see this
Re-request review when you finish

Glasis9 committed:
I did it!
Reduced the time from 59 seconds to 9.7 seconds (quote + authors)
Used a combination of threads and processes

Glasis9 commented:
![quote + author + threding and processing](https://user-images.githubusercontent.com/108976083/219469933-7ea3c81a-2772-4981-acae-62d37ac76c19.jpg)
![authors](https://user-images.githubusercontent.com/108976083/219469943-2349151a-8bb4-4edd-9688-2bbc8504e206.jpg)
![quotes](https://user-images.githubusercontent.com/108976083/219469954-19d99dc2-3f35-416e-9190-0f55aed9677e.jpg)

Glasis9 committed:
rewrote logic implementation threading

Glasis9 committed:
fix flake8

ArsenPidhoretskyi reviewed (approved):
Good job!
